### DISCUSSION TITLE
Adding fillets or chamfers to an SVG extrusion

### DISCUSSION THREAD

**Discussion by gautaz** [gumyr/build123d#810] [Q&A] [answered]
Hello,

Lately I've stumbled upon an issue I haven't yet found a workaround for and I'd like to know if this issue is related to my environment or if this is something inherent to the way build123d and opencascade operate.

As described in #805 I am trying to add fillets to the extrusion of an SVG path and this currently fails with this kind of message:

```console
Traceback (most recent call last):
  File "/models/./blob.py", line 20, in <module>
    fillet(blob.edges().group_by(Axis.Z)[-1], radius=1)
  File "/opt/build123d/lib/python3.12/site-packages/build123d/operations_generic.py", line 426, in fillet
    new_part = target.fillet(radius, list(object_list))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/build123d/lib/python3.12/site-packages/build123d/topology.py", line 991, in fillet
    fillet_builder.Add(radius, native_edge)
OCP.gp.gp_VectorWithNullMagnitude:
```

I have tried with different SVG paths, fillet/chamfer radius, edge selections, it doesn't seem to help.

So has anybody stumbled upon this `VectorWithNullMagnitude` issue?
I understand that there is somewhere a vector with a zero length so it does not provide any direction to operate some computation depending on it, but which vector are we talking about?
A vector normal to the fillet/chamfer?
Something else?

**Comment by lypwig** [comment]
If you create for instance a 2cm chamfer on 4cm wide face, the produced face will be 0 cm wide and OCP doesn't like that. You can check if it works with smaller chamfer.

Sometimes I use an epsilon variable to avoid this problem:

```py
EPSILON = 0.01
# ...
box = _.chamfer(box.edges()[0], chamfer_size - EPSILON)
```

**Reply by gautaz** [reply]
Thanks for the tip, I'll keep that in mind :+1:

**Comment by gumyr** [comment]
As you can see working with SVG files can be quite difficult.  Is there a limitation to the build123d line drawing capabilities that is causing you to go to SVG?

**Reply by gautaz** [reply]
Hello @gumyr,

I was using an SVG because I wanted to easily draw the shape of a cat hence the use of Gimp to cut down a cat image and produce the SVG I was originally trying to extrude and fillet.

So the issue here is not related to build123d drawing capabilities.
They really shine when one wants to define parametrically a shape from scratch.
This is in fact what I am doing most of the time, but not for this particular "cat" project.

Your question is interesting because now I am wondering if I can use the SVG content and transform it into build123d native instructions as all instructions provided by SVG (line, circle, bezier, ...) are also available directly in build123d.

Perhaps there is still hope for "filleting the cat" :slightly_smiling_face:.

**Reply by gumyr** [reply, marked as answer]
Try https://build123d.readthedocs.io/en/latest/import_export.html#importers.import_svg_as_buildline_code. It will generate the first draft of the code that you can edit to remove the weirdness.

**Reply by gautaz** [reply]
Wow ! I didn't realize It's already off the shelf, now I feel dumb :rofl:.
Thanks, I'll try this and keep you posted :+1:.

**Reply by gumyr** [reply]
What's there is quite simple and could be enhanced with features like:

- Automatically rounding off values to a given precision
- Identifying circles and not always using an ellipse
- Use the `l @ 0` and `l @ 1` operator to link objects together - basically identify common points like the end of one line being used as the beginning of another
- Being able to flip along the Y Axis

I created this because I had SVG import problems like you're experiencing.

**Reply by gautaz** [reply]
At last a success!
I did manage to fillet the cat :slightly_smiling_face:
Thanks a lot @gumyr for the `import_svg_as_buildline_code` function and all the advices.

![image](https://github.com/user-attachments/assets/942450f3-a774-484e-8b52-189b6c3e373f)

The code is available [here](https://github.com/gautaz/cadcode/blob/main/models/reading-ring.py).

A little feedback from my personal experience:

> Automatically rounding off values to a given precision

Careful with rounding the `9999999` occurrences, I tried to manually round some values as an experiment:

```python
            Bezier((99.62, 483.0),(103.7, 475.55),(97.7, 462.47),(96.28, 454.0))
            Bezier((96.28, 454.0),(93.36, 436.52),(94.2, 438.75),(94.0, 422.0))
            Bezier((94.0, 422.0),(93.8, 406.03),(80.07, 397.01),(81.04, 380.0))
            Bezier((81.04, 380.0),(81.98, 363.62),(106.39466, 370.23644),(113.0, 365.59))
            Bezier((113.0, 365.59),(160.41358, 325.85073),(196.40241, 340.63727),(211.0, 336.56))
            Bezier((211.0, 336.56),(221.48644, 333.01644),(232.62, 316.83),(244.99, 323.17))
            Bezier((244.99, 323.17),(255.15, 328.39),(254.87, 345.19),(258.03, 355.0))
            Bezier((258.03, 355.0),(260.49, 362.65),(263.18, 364.48),(266.68, 371.0))
            Bezier((266.68, 371.0),(269.11, 375.53),(273.06, 393.19),(294.0, 394.82))
            Bezier((294.0, 394.82),(354.60047, 399.42444),(400.04966, 350.77694),(507.0, 366.71))
            Bezier((507.0, 366.71),(533.49496, 368.12847),(534.17535, 342.72412),(582.0, 344.0))
            Bezier((582.0, 344.0),(625.89, 343.68),(666.21, 358.48),(706.0, 358.0))
            Bezier((706.0, 358.0),(724.49585, 357.73876),(738.35758, 335.30056),(752.15, 351.19))
            Bezier((752.15, 351.19),(760.697, 361.03528),(755.57774, 374.26173),(753.57, 379.0))
            Bezier((753.57, 379.0),(742.85976, 401.51133),(710.12811, 406.51489),(687.0, 405.96))
            Bezier((687.0, 405.96),(675.68381, 405.4277),(658.38095, 396.81077),(653.25, 412.0))
            Bezier((653.25, 412.0),(651.62303, 426.66525),(668.45823, 432.59708),(679.0, 439.06))
            Bezier((679.0, 439.06),(733.13489, 472.68876),(743.39365, 518.91545),(693.0, 519.0))
            Bezier((693.0, 519.0),(660.64109, 518.60725),(628.32132, 513.47212),(596.0, 517.29))
            Bezier((596.0, 517.29),(518.11547, 527.78021),(474.18646, 600.65979),(369.0, 600.0))
            Bezier((369.0, 600.0),(264.41304, 590.48259),(226.71723, 556.40199),(145.0, 561.0))
            Bezier((145.0, 561.0),(113.26242, 560.50478),(34.658453, 589.82408),(37.04, 540.0))
            Bezier((37.04, 540.0),(39.652887, 506.6985),(86.892464, 506.29687),(99.62, 483.0))
```

This failed with this error message:

```console
build123d-1  | 2024-12-21T14:48:42+01:00 INFO:root:Exporting...
build123d-1  | 2024-12-21T14:50:25+01:00 Traceback (most recent call last):
build123d-1  | 2024-12-21T14:50:25+01:00   File "/models/./reading-ring.py", line 73, in <module>
build123d-1  | 2024-12-21T14:50:25+01:00     exporter3d.add_shape(cat.part)
build123d-1  | 2024-12-21T14:50:25+01:00   File "/opt/build123d/lib/python3.12/site-packages/build123d/mesher.py", line 432, in add_shape
build123d-1  | 2024-12-21T14:50:25+01:00     raise RuntimeError("3mf mesh is invalid")
build123d-1  | 2024-12-21T14:50:25+01:00 RuntimeError: 3mf mesh is invalid
```

As for the SVG itself, I processed it this way:

1. I removed all control points that were not really necessary
2. I edited (with inkscape) the control points where tangents were too "unparallel" (i.e. creating local corners)

This:
![image](https://github.com/user-attachments/assets/eb7c1d84-d88c-47d4-ab07-9c103ce7f449)

was replaced with that:
![image](https://github.com/user-attachments/assets/bbba7d39-ba8c-43c3-a115-f7d9a085fb54)

**Reply by gumyr** [reply]
Congratulations, that's a nice project.  With respect to the rounding, you're correct the points need to match up.  The first step to improve this should be:
```py
            l1 = Bezier((99.62, 483.0),(103.7, 475.55),(97.7, 462.47),(96.28, 454.0))
            l2 = Bezier(l1@1,(93.36, 436.52),(94.2, 438.75),(94.0, 422.0))
            l3 = Bezier(l2@1,(93.8, 406.03),(80.07, 397.01),(81.04, 380.0))
            ...
```
where the end of one line is explicitly used as the start of the next (a very common pattern in these drawings).  This ensures that they automatically connect up even if there is a little change caused by rounding.

If you do a similar project in the future you might want to consider creating the BuildLine code directly.  The ocp_vscode viewer supports importing a picture which can be used as a guide as you create the CAD objects in place.  Depending on the SVG is might be faster.
